lib/rex/proto/ldap/auth.rb

```diff
@@ -0,0 +1,348 @@
+require 'net/ldap'
+require 'net/ldap/dn'
+
+module Rex
+  module Proto
+    module LDAP
+      class Auth
+        NTLM_CONST = Rex::Proto::NTLM::Constants
+        NTLM_CRYPT = Rex::Proto::NTLM::Crypt
+        MESSAGE = Rex::Proto::NTLM::Message
+
+        #
+        # Initialize the required variables
+        #
+        # @param opts [Hash] Required authentication options
+        def initialize(opts = {}) # chech for the options recieved
+          @domain = opts[:domain]
+          @server = opts[:server]
+          @dnsname = opts[:dnsname]
+          @dnsdomain = opts[:dnsdomain]
```

Comment:
Why are these values being stored in the `Auth` instance?  They should be extracted from the Auth process for all current use cases.  The only reason to have them in a class here would be if the Authentication handler needs to know to reject non-matching values.

Also by not checking values this could cause runtime issues if the class instance values were not valid.

Reply:
I do like the idea of requiring the configuration of the `Auth` class in the constructor and all values being private to the object, however only values that cannot be obtained during the auth request should be part of the constructor.

Reply:
These values are set on the server and are required to generate type2 response. These are not extracted from the `PDU` object

Reply:
With these values injected there needs to be validation guards around `nil` values, this is one reason I prefer `named parameters` or parameters with default values in constructors in the `Rex`  namespace.

Validation could be handled either here in this class or encapsulated in a class for `auth_options` that preforms the validation on initialize allowing this class to just ensure `auth_options` is of the class type or simply create an instance of class internally to be responsible for that validation.

Some more thoughts based on looking at this in isolation:

The existing module attempts to extract `domain` and `server` from the type1 message and falls back if the values are not in the request. I do see that the datastore `dnsname` and `dnsdomain` values are currently always used from the module options.

I could see using these values as a filtering method, I also am interested in if the values impact the NTLM challenge process in a real auth attempt. Would the client reject a challenge based on unexpected values?

There is a need to work out some testing patten here to validate expected behavior when offered an NTLM bind request that does not match the values set as that could both provide a signature for a rouge service as well as limit the usability of the capture services when NTLM auth occurs in a target network.

In future iterations this may also impact how kerberos captures need to be processed although if offered a kerberos ticket I am fairly certain all these details can simply be extracted from it directly.